dali/pipeline/operators/arithmetic/arithmetic_meta.h

```diff
@@ -0,0 +1,345 @@
+  REGISTER_UNARY_IMPL_BACKEND(OP, EXPRESSION, CPUBackend) \
+  REGISTER_UNARY_IMPL_BACKEND(OP, EXPRESSION, GPUBackend)
+
+REGISTER_UNARY_IMPL(ArithmeticOp::plus, +);
+REGISTER_UNARY_IMPL(ArithmeticOp::minus, -);
+
+#define REGISTER_BINARY_IMPL_BACKEND(OP, EXPRESSION, BACKEND)                          \
+  template <>                                                                          \
+  template <typename L, typename R>                                                    \
+  DALI_HOST_DEV constexpr std::enable_if_t<GetOpArity(OP) == 2, binary_result_t<L, R>> \
+  arithm_meta<OP, BACKEND>::impl(L l, R r) {                                           \
+    static_assert(GetOpArity(OP) == 2,                                                 \
+                  "Registered operation arrity does not match the requirements.");     \
+    return l EXPRESSION r;                                                             \
+  }                                                                                    \
+  template <>                                                                          \
+  inline std::string arithm_meta<OP, BACKEND>::to_string() {                           \
+    return #EXPRESSION;                                                                \
+  }
+
+#define REGISTER_BINARY_IMPL(OP, EXPRESSION)               \
+  REGISTER_BINARY_IMPL_BACKEND(OP, EXPRESSION, CPUBackend) \
+  REGISTER_BINARY_IMPL_BACKEND(OP, EXPRESSION, GPUBackend)
+
+REGISTER_BINARY_IMPL(ArithmeticOp::add, +);
+REGISTER_BINARY_IMPL(ArithmeticOp::sub, -);
+REGISTER_BINARY_IMPL(ArithmeticOp::mul, *);
+REGISTER_BINARY_IMPL(ArithmeticOp::div, /);
+
+template <>
+struct arithm_meta<ArithmeticOp::mod, CPUBackend> {
+  template <typename L, typename R>
+  static constexpr std::enable_if_t<
+      std::is_integral<L>::value && std::is_integral<R>::value, binary_result_t<L, R>>
+  impl(L l, R r) {
+    return l % r;
+  }
+
+  template <typename L, typename R>
+  static constexpr std::enable_if_t<
+      !std::is_integral<L>::value || !std::is_integral<R>::value, binary_result_t<L, R>>
+  impl(L l, R r) {
+    using L_promotion = std::conditional_t<std::is_same<float16, L>::value, float, L>;
+    using R_promotion = std::conditional_t<std::is_same<float16, R>::value, float, R>;
+    return std::remainder(static_cast<L_promotion>(l), static_cast<R_promotion>(r));
+  }
+
+  static std::string to_string() {
+    return "%";
+  }
+
+  static constexpr int num_inputs = 2;
+  static constexpr int num_outputs = 1;
+};
+
+template <>
+struct arithm_meta<ArithmeticOp::mod, GPUBackend> {
+  template <typename L, typename R>
+  __device__ static constexpr std::enable_if_t<
+      std::is_integral<L>::value && std::is_integral<R>::value, binary_result_t<L, R>>
+  impl(L l, R r) {
+    return l % r;
+  }
+
+  template <typename L, typename R>
+  __device__ static constexpr std::enable_if_t<
+      (!std::is_integral<L>::value || !std::is_integral<R>::value) &&
```

Comment:
note: In `float16.h` you have this:
```
template <typename T>
struct is_fp_or_half {
  static constexpr bool value =
    std::is_floating_point<T>::value || is_half<T>::value;
};
```
Might it be useful to detect the floating-point case, or do you prefer to define it for anything that is not integral?

Reply:
Might be useful, but only integers work with `%`. Still, it's probably going to be changed, we need to decide if we follow c++ or python way.